File: ios/RNAdManagerBannerAdCache.m
// RNAdManagerBannerAdCache.m

#import "RNAdManagerBannerAdCache.h"

@implementation RNAdManagerBannerAdCache {
    NSMutableDictionary<NSString *, GADBannerView *> *_bannerAdCache;
}

RCT_EXPORT_MODULE(CTKAdManagerBannerAdCache)

+ (instancetype)sharedInstance {
    static RNAdManagerBannerAdCache *sharedInstance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedInstance = [[self alloc] init];
    });
    return sharedInstance;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        _bannerAdCache = [NSMutableDictionary dictionary];
    }
    return self;
}

- (void)cacheBannerView:(GADBannerView *)bannerView
            forAdUnitID:(NSString *)adUnitID
       selectedCategory:(NSString *)selectedCategory
               adAtIndex:(NSString *)adIndex {
    if (bannerView && adUnitID && selectedCategory && adIndex) {
        NSString *resultKey = [self cacheKeyForAdUnitID:adUnitID selectedCategory:selectedCategory adAtIndex:adIndex];
        _bannerAdCache[resultKey] = bannerView;
        NSLog(@"Stored bannerView: %@ for adUnitID: %@ of selectedCategory: %@  adAtIndex: %@", bannerView, adUnitID, selectedCategory, adIndex);
    }
}

- (GADBannerView *)cachedBannerViewForAdUnitID:(NSString *)adUnitID
                                      selectedCategory:(NSString *)selectedCategory
                                              adAtIndex:(NSString *)adIndex {
    if (adUnitID && selectedCategory && adIndex) {
        NSString *resultKey = [self cacheKeyForAdUnitID:adUnitID selectedCategory:selectedCategory adAtIndex:adIndex];
        GADBannerView *cachedBannerView = _bannerAdCache[resultKey];
        if (cachedBannerView) {
            NSLog(@"Successfully fetched cachedBannerView: %@ for adUnitID: %@ of selectedCategory: %@  adAtIndex: %@", cachedBannerView, adUnitID, selectedCategory, adIndex);
            return cachedBannerView;
        }
    }
    return NULL;
}

- (NSString *)cacheKeyForAdUnitID:(NSString *)adUnitID
                selectedCategory:(NSString *)selectedCategory
                        adAtIndex:(NSString *)adIndex {
    NSString *cacheKey = [NSString stringWithFormat:@"%@_%@_%@", adUnitID, selectedCategory, adIndex];
    return cacheKey;
}

- (void)clearCache {
    _bannerAdCache = [NSMutableDictionary dictionary];
    NSLog(@"Cleared cache Ad data");
}

RCT_EXPORT_METHOD(clearStoredCache) {
    [[RNAdManagerBannerAdCache sharedInstance] clearCache];
}


@end

File: android/src/main/java/com/matejdr/admanager/RNAdManagerBannerAdCache.java
package com.matejdr.admanager;

import android.util.Log;
import com.google.android.gms.ads.admanager.AdManagerAdView;

import com.facebook.react.module.annotations.ReactModule;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import java.util.HashMap;
import java.util.Map;

@ReactModule(name = "CTKAdManagerBannerAdCache")
public class RNAdManagerBannerAdCache extends ReactContextBaseJavaModule {

    @Override
    public String getName() {
        return REACT_CLASS;
    }

    public static final String REACT_CLASS = "CTKAdManagerBannerAdCache";

    private static RNAdManagerBannerAdCache sharedInstance;
    private Map<String, AdManagerAdView> bannerAdCache;

    private RNAdManagerBannerAdCache() {
        bannerAdCache = new HashMap<>();
    }

    public static RNAdManagerBannerAdCache getSharedInstance() {
        if (sharedInstance == null) {
            synchronized (RNAdManagerBannerAdCache.class) {
                if (sharedInstance == null) {
                    sharedInstance = new RNAdManagerBannerAdCache();
                }
            }
        }
        return sharedInstance;
    }

    public void cacheBannerView(AdManagerAdView bannerView, String adUnitID, String selectedCategory, String adIndex) {
        if (bannerView != null && adUnitID != null && selectedCategory != null && adIndex != null) {
            String resultKey = cacheKeyForAdUnitID(adUnitID, selectedCategory, adIndex);
            bannerAdCache.put(resultKey, bannerView);
            Log.d("RNAdManagerBannerAdCache", "Stored bannerView: " + bannerView + " for adUnitID: " + adUnitID +
                    " of selectedCategory: " + selectedCategory + " adAtIndex: " + adIndex);
        }
    }

    public AdManagerAdView cachedBannerViewForAdUnitID(String adUnitID, String selectedCategory, String adIndex) {
        if (adUnitID != null && selectedCategory != null && adIndex != null) {
            String resultKey = cacheKeyForAdUnitID(adUnitID, selectedCategory, adIndex);
            AdManagerAdView cachedBannerView = bannerAdCache.get(resultKey);
            if (cachedBannerView != null) {
                Log.d("RNAdManagerBannerAdCache", "Successfully fetched cachedBannerView: " + cachedBannerView +
                        " for adUnitID: " + adUnitID + " of selectedCategory: " + selectedCategory +
                        " adAtIndex: " + adIndex);
                return cachedBannerView;
            }
        }
        return null;
    }

    private String cacheKeyForAdUnitID(String adUnitID, String selectedCategory, String adIndex) {
        return adUnitID + "_" + selectedCategory + "_" + adIndex;
    }

    @ReactMethod
    public void clearCache() {
        bannerAdCache.clear();
        Log.d("RNAdManagerBannerAdCache", "Cleared cache Ad data");
    }

    public void clearStoredCache() {
        getSharedInstance().clearCache();
    }
}

